autorest/azure/token.go

```diff
@@ -104,6 +154,79 @@ func (spt *ServicePrincipalToken) EnsureFresh() error {
 
 // Refresh obtains a fresh token for the Service Principal.
 func (spt *ServicePrincipalToken) Refresh() error {
+	newToken, err := spt.GetOauthToken()
+	if err != nil {
+		return err
+	}
+	spt.Token = *newToken
+
+	return nil
+}
+
+// SetAuthValues is a method of the interface ServicePrincipalTokenSecret
+// It will populate the form submitted during oAuth Token Acquisition using a static client_secret
+func (tokenSecret *ServicePrincipalTokenSecret) SetAuthValues(spt *ServicePrincipalToken, v *url.Values) error {
+	v.Set("client_secret", tokenSecret.ClientSecret)
+	return nil
+}
+
+// SignedJwt returns the JWT signed with the secret's certData
+func (secret *ServicePrincipalCertSecret) SignedJwt(spt *ServicePrincipalToken) (string, error) {
+	privateKey, cert, err := pkcs12.Decode(secret.Pkcs12Bytes, secret.Password)
+	if err != nil {
+		return "", err
+	}
+
+	rsaPrivKey, isRsaKey := privateKey.(*rsa.PrivateKey)
+	if !isRsaKey {
+		return "", fmt.Errorf("PKCS12 certificate must contain an RSA private key")
+	}
+
+	hasher := sha1.New()
+	_, err = hasher.Write(cert.Raw)
+	if err != nil {
+		return "", err
+	}
+
+	thumbprint := base64.URLEncoding.EncodeToString(hasher.Sum(nil))
+
+	jti := make([]byte, 20)
+	_, err = rand.Read(jti)
+	if err != nil {
+		return "", err
+	}
+
+	token := jwt.New(jwt.SigningMethodRS256)
+	token.Header["x5t"] = thumbprint
+	token.Claims["aud"] = fmt.Sprintf("https://login.microsoftonline.com/%s/oauth2/token", spt.tenantID)
```

Comment:
Please define a constant to hold the format string. Burying these in code makes finding / fixing / changing them very hard.